### PR TITLE
The great UTF8 Refactoring

### PR DESCRIPTION
**Pull request by vmg**
Here it is. The long awaited UTF8 compatibility refactoring. This new autolinking implementation is now UTF8 aware, and should handle all previously reported encoding-related bugs and corner cases.

It also includes fixes for mixed encodings, better handling of strings with bad encodings, and fixes for small autolinking bugs (like hash-encoded unicode entities at the end of a link, or case-sensitive `WWW` prefixes).

I've backported all of @kerrizor's tests and written a bunch of new ones. 👌 

Fixes https://github.com/vmg/rinku/issues/50
Fixes https://github.com/vmg/rinku/issues/39

Fixes https://github.com/vmg/rinku/pull/51
Fixes https://github.com/vmg/rinku/pull/43
Fixes https://github.com/vmg/rinku/pull/42
Fixes https://github.com/vmg/rinku/pull/28
Fixes https://github.com/vmg/rinku/pull/44

### CODE REVIEW

**Comment by grosser** [comment]
😍

**Comment by magnusvk** [comment]
Does the 2.0.0 gem release contain these changes?

**Comment by magnusvk** [comment]
It appears so; it fixes our UTF issue, at least. 😄 Thanks for the update.

**Comment by vmg** [comment]
You're welcome. The 2.0.0 release does indeed contain all these changes, and fixes all the known UTF issues. Please report anything I might have missed! :)

File: test/autolink_test.rb [hunk 1]
```diff
@@ -279,23 +286,94 @@ def test_links_like_autolink_rails
...
+    str = "&lt;a href=&#39;http://gi.co&#39;&gt;gi.co&lt;/a&gt;\xC2\xA0r"
+    assert_equal Encoding::UTF_8, str.encoding
+
+    res = Rinku.auto_link(str)
+    assert_equal Encoding::UTF_8, res.encoding
+    assert res.valid_encoding?
+  end
+
+  def test_polish_wikipedia_haha
+    url = "https://pl.wikipedia.org/wiki/Komisja_śledcza_do_zbadania_sprawy_zarzutu_nielegalnego_wywierania_wpływu_na_funkcjonariuszy_policji,_służb_specjalnych,_prokuratorów_i_osoby_pełniące_funkcje_w_organach_wymiaru_sprawiedliwości"
```
Comment (Holek) [inline test/autolink_test.rb]: ❤️

Comment (vmg) [inline test/autolink_test.rb]: I tried to pick up every substantial test case from the open issues. Thanks for reporting! 😄

**Pull request merged**